File: algorithms/solutions/week26/manhdv/binary_tree_level_order_traversal.rb
# https://leetcode.com/problems/binary-tree-level-order-traversal/description/

def trace_node(node, level, result)
  return if node.nil?

  result[level] ||= []
  result[level] << node.val unless node.val.nil?

  trace_node(node.left, level + 1, result) unless node.left.nil?
  trace_node(node.right, level + 1, result) unless node.right.nil?
end

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[][]}
def level_order(root)
  result = {}

  trace_node(root, 0, result)

  result.values
end

File: algorithms/solutions/week26/toanha/binary-tree-level-order-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func appendAtLevel(root *TreeNode, arr [][]int, level int) [][]int {
	if root == nil {
		return arr
	}

	if len(arr) == level {
		arr = append(arr, []int{})
	}

	arr[level] = append(arr[level], root.Val)
	arr = appendAtLevel(root.Left, arr, level+1)

	return appendAtLevel(root.Right, arr, level+1)
}

func levelOrder(root *TreeNode) [][]int {
	return appendAtLevel(root, [][]int{}, 0)
}
